manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,244 @@
+package deallocator
+
+import (
+	"reflect"
+
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+	"golang.org/x/net/context"
+)
+
+var (
+	resourceDeallocators = make(map[reflect.Type]resourceDeallocator)
+)
+
+func registerServiceLevelResource(resourceDeallocator resourceDeallocator) {
+	resourceDeallocators[reflect.TypeOf(resourceDeallocator.event)] = resourceDeallocator
+}
+
+// the Deallocator waits for services to fully shutdown (ie no containers left)
+// and then proceeds to deallocate service-level resources (e.g. networks),
+// and finally services themselves
+type Deallocator struct {
+	store *store.MemoryStore
+
+	// for services that are shutting down, we keep track of how many
+	// tasks still exist for them
+	services map[string]serviceWithTaskCounts
+
+	stopChan chan struct{}
+	doneChan chan struct{}
+}
+
+type serviceWithTaskCounts struct {
+	service   *api.Service
+	taskCount int
+}
+
+// New creates a new deallocator
+func New(store *store.MemoryStore) *Deallocator {
+	return &Deallocator{
+		store:    store,
+		services: make(map[string]serviceWithTaskCounts),
+
+		stopChan: make(chan struct{}),
+		doneChan: make(chan struct{}),
+	}
+}
+
+// Run starts the deallocator, which then starts cleaning up services
+// and their resources when relevant (ie when no tasks still exist
+// for a given service)
+// This is a blocking function
+func (deallocator *Deallocator) Run(ctx context.Context) error {
+	// those are the events we'll be watching
+	events := make([]api.Event, len(resourceDeallocators)+2)
+	// we want to know when tasks are deleted by the reaper
+	events[0] = api.EventDeleteTask{}
+	// and when services get marked for deletion
+	events[1] = api.EventUpdateService{Service: &api.Service{PendingDelete: true}}
+	i := 2
+	for _, resourceDeallocator := range resourceDeallocators {
+		events[i] = resourceDeallocator.event
```

Comment:
Can you please clarify why we need to do this?

Reply:
I'm not 100% sure I understand the question correctly, but basically this also boils down to separating the resource-specific code from the core of the deallocator: each resource-specific file defines the API event it cares about, and this is just aggregating all that to feed it to `store.ViewAndWatch` below.

Please let me know if I misunderstood the question!